### DISCUSSION TITLE
Report the responding server in a custom header

### DISCUSSION THREAD

**Discussion by rootix** [dotnet/yarp#1203] [Q&A] [answered]
Hi,

I want to return the responding server in a load balanced configuration back to the caller to see from which destination the response was generated. This is only used within a dev environment. I configured a single route with a single cluster and wrote this code, which works for this simple setup:

```cs
services.AddReverseProxy()
    .LoadFromConfig(this.configuration.GetSection("ReverseProxy"))
    .AddTransforms(context =>
    {
        context.AddResponseTransform(ctx =>
	{
            /* Compares ctx.ProxyResponse.RequestMessage.RequestUri with the Cluster destination adresses to find the 
               responding one and return its key */
	    var responseServer = DetectResponseServer(ctx, context.Cluster);
	    if (!string.IsNullOrEmpty(responseServer))
	    {
	        ctx.HttpContext.Response.Headers.Add("yarp-response-server", responseServer);
	    }
	
            return default;
	});
    });
```

The problem with this code is, that the context variable from `AddTransforms` is captured and reused for every subsequent call. I suspect it will contain the resolved cluster from the first call.

Is there a way of doing something like this? Looking at the load balancer middleware, the decision is not stored and the `ResponseTransformContext` does not hold any information either. I guess I could check the raw `IConfiguration` for all clusters and get it from there, but if the same server is used in multiple clusters (is there an use case for this?), it could report the wrong name.

**Comment by MihaZupan** [comment, marked as answer]
The YARP pipeline will add an `IReverseProxyFeature` to the `HttpContext`. The feature also stores the destination that was chosen for proxying by health checks and load balancing.

```c#
var responseServer = ctx.HttpContext.Features.Get<IReverseProxyFeature>()?.ProxiedDestination;
if (responseServer is not null)
{
    ctx.HttpContext.Response.Headers.Add("yarp-response-server", responseServer.DestinationId);
}
```

**Reply by rootix** [reply]
Thanks, this is awesome. It even works for static files, which my code didn't do somehow.

I will mark the answer as soon as Github let me do it :)
![image](https://user-images.githubusercontent.com/874015/130909287-e5ef87a4-0370-4fee-8ad5-9c5c228fb835.png)
